File: backend/app/services/lottery_service.py
from datetime import datetime
from typing import List, Dict, Tuple
from sqlalchemy.orm import Session
from sqlalchemy import select, func

from app.models.lottery import LotteryDraw, LotteryType
from app.services.cache import cache
from app.services.fetchers import fetch_hk_recent, fetch_macau_recent

CACHE_SECONDS = 600


def normalize_numbers(numbers: str) -> str:
	# Ensure numbers are in comma-separated format without spaces
	parts = [p.strip() for p in numbers.replace(" ", ",").replace("|", ",").split(",") if p.strip()]
	return ",".join(parts)


async def get_recent_draws(db: Session, lottery_type: LotteryType, limit: int = 10) -> List[Dict]:
	cache_key = f"recent:{lottery_type.value}:{limit}"
	cached = cache.get_json(cache_key)
	if cached:
		return cached

	items: List[Dict]
	if lottery_type == LotteryType.HK:
		items = await fetch_hk_recent(limit=limit)
	elif lottery_type == LotteryType.MACAU:
		items = await fetch_macau_recent(limit=limit)
	else:
		# From DB for NEW_HK
		stmt = select(LotteryDraw).where(LotteryDraw.lottery_type == lottery_type).order_by(LotteryDraw.open_time.desc()).limit(limit)
		rows = db.execute(stmt).scalars().all()
		items = [
			{
				"id": row.id,
				"lottery_type": row.lottery_type.value,
				"issue": row.issue,
				"numbers": row.numbers,
				"open_time": row.open_time.isoformat(sep=" ")
			}
			for row in rows
		]
		cache.set_json(cache_key, items, ex_seconds=CACHE_SECONDS)
		return items

	# Persist external draws to DB (upsert by unique issue)
	for item in items:
		issue = item.get("issue") or ""
		numbers = normalize_numbers(item.get("numbers") or "")
		open_time_str = item.get("open_time") or ""
		open_time = datetime.fromisoformat(open_time_str.replace("Z", "").replace("T", " ")) if open_time_str else datetime.utcnow()
		existing = db.execute(
			select(LotteryDraw).where(
				LotteryDraw.lottery_type == lottery_type,
				LotteryDraw.issue == str(issue),
			)
		).scalar_one_or_none()
		if existing is None:
			row = LotteryDraw(lottery_type=lottery_type, issue=str(issue), numbers=numbers, open_time=open_time)
			db.add(row)
	try:
		db.commit()
	except Exception:
		db.rollback()

	cache.set_json(cache_key, items, ex_seconds=CACHE_SECONDS)
	return items


def get_paginated_draws(db: Session, lottery_type: LotteryType, page: int, size: int) -> Tuple[List[LotteryDraw], int]:
	offset = (page - 1) * size
	count_stmt = select(func.count()).select_from(LotteryDraw).where(LotteryDraw.lottery_type == lottery_type)
	total = db.execute(count_stmt).scalar_one()
	stmt = (
		select(LotteryDraw)
		.where(LotteryDraw.lottery_type == lottery_type)
		.order_by(LotteryDraw.open_time.desc())
		.offset(offset)
		.limit(size)
	)
	rows = db.execute(stmt).scalars().all()
	return rows, int(total)


def create_new_hk_draw(db: Session, issue: str, numbers: str, open_time: datetime) -> LotteryDraw:
	numbers_norm = normalize_numbers(numbers)
	draw = LotteryDraw(lottery_type=LotteryType.NEW_HK, issue=str(issue), numbers=numbers_norm, open_time=open_time)
	db.add(draw)
	db.commit()
	db.refresh(draw)
	# Invalidate caches
	cache.set_json("recent:NEW_HK:10", None, ex_seconds=1)
	return draw

File: node-backend/src/services/lotteryService.js
import { Op } from 'sequelize'
import { sequelize, defineModels } from '../db.js'
import { getJson, setJson } from './cache.js'
import { fetchHKRecent, fetchMacauRecent } from './fetchers.js'

const { LotteryDraw } = defineModels()
const CACHE_SECONDS = 600

function normalizeNumbers(numbers) {
	return numbers.replace(/\s+/g, ',').replace(/\|/g, ',').split(',').map(s => s.trim()).filter(Boolean).join(',')
}

export async function getRecentDraws(lotteryType, limit = 10) {
	const key = `recent:${lotteryType}:${limit}`
	const cached = await getJson(key)
	if (cached) return cached
	let items
	if (lotteryType === 'HK') items = await fetchHKRecent(limit)
	else if (lotteryType === 'MACAU') items = await fetchMacauRecent(limit)
	else {
		const rows = await LotteryDraw.findAll({ where: { lottery_type: 'NEW_HK' }, order: [['open_time','DESC']], limit })
		items = rows.map(r => ({ id: r.id, lottery_type: r.lottery_type, issue: r.issue, numbers: r.numbers, open_time: r.open_time.toISOString() }))
		await setJson(key, items, CACHE_SECONDS)
		return items
	}
	// upsert into DB
	for (const item of items) {
		const issue = String(item.issue || '')
		const numbers = normalizeNumbers(item.numbers || '')
		const openTime = new Date(item.open_time)
		await LotteryDraw.findOrCreate({
			where: { lottery_type: lotteryType, issue },
			defaults: { lottery_type: lotteryType, issue, numbers, open_time: openTime }
		})
	}
	await setJson(key, items, CACHE_SECONDS)
	return items
}

export async function getPaginatedDraws(lotteryType, page = 1, size = 10) {
	const offset = (page - 1) * size
	const { rows, count } = await LotteryDraw.findAndCountAll({
		where: { lottery_type: lotteryType },
		order: [['open_time','DESC']],
		limit: size,
		offset,
	})
	return { items: rows, total: count }
}

export async function createNewHK(issue, numbers, openTimeStr) {
	const open_time = new Date(openTimeStr)
	const numbersNorm = normalizeNumbers(numbers)
	const row = await LotteryDraw.create({ lottery_type: 'NEW_HK', issue: String(issue), numbers: numbersNorm, open_time })
	// invalidate cache by setting short ttl
	await setJson('recent:NEW_HK:10', null, 1)
	return row
}